src/features/machine/selectors.js

```diff
@@ -1,47 +1,98 @@
 import { createSelector } from 'reselect'
 import {
   transformShapes,
-  patternImport
+  transformShape,
+  polishVertices,
+  scaleImportedVertices
 } from './computer'
+import { getShape } from '../shapes/selectors'
 
 const getApp = state => state.app
 const getShapes = state => state.shapes
 const getTransforms = state => state.transforms
+const getCurrentShape = state => state.shapes.byId[state.shapes.currentId]
+const getTransform = state => state.transforms.byId[state.shapes.currentId]
 const getImporter = state => state.importer
-const getExporter = state => state.exporter
 const getMachine = state => state.machine
 
+// by returning null for shapes which can change size, this selector will ensure
+// transformed vertices are not redrawn when machine settings change
+const getTransformMachine = state => getTransform(state).canChangeSize ? null : state.machine
+
+// requires only shape, and in the case of erasers, machine state
+export const getShapedVertices = createSelector(
```

Comment:
These selectors are the key then? If you cascade them with just the minimum dependencies, it doesn't require all the calculations to happen again? And ciritically, the getCurrentShape is only dependent on the shape items, not the lower down selectors?

Reply:
Yes, exactly.